File: Measure steerability utilizing ANNs/Multipartite steering scenarios/3-qubit_2-unt/targets.py
import numpy as np
import tensorflow as tf

# Function to generate GHZ assemblage **(2-UNT & 2-M)**
def GHZ_assemblage_2unt_2m(w):
    """
    TThe assemblage created for Charlie after Alice and Bob each performs two Pauli spin measurements
    (X and Z) on her/his particle of the three-qubit GHZ state.
    """
    Pauli_x = np.array([[0, 1], [1, 0]])
    Pauli_z = np.array([[1, 0], [0, -1]])
    sigma = np.zeros((2, 2, 2, 2, 2, 2), dtype=complex)
    diagonal = np.diag([(1 - w) / 8, (1 - w) / 8])
    Pauli = [Pauli_x, Pauli_z]
    for x in range(2):
        for a in range(2):
            for y in range(2):
                for b in range(2):
                    projector_A = (np.eye(2) + ((-1) ** a) * Pauli[x]) / 2
                    projector_B = (np.eye(2) + ((-1) ** b) * Pauli[y]) / 2
                    sigma[:, :, a, b, x, y] = diagonal + (w / 2) * np.array([
                        [projector_A[0, 0] * projector_B[0, 0], projector_A[1, 0] * projector_B[1, 0]],
                        [projector_A[0, 1] * projector_B[0, 1], projector_A[1, 1] * projector_B[1, 1]]
                    ])
    return tf.cast(sigma,'complex64')

# Function to generate GHZ assemblage **(2-UNT & 3-M)**
def GHZ_assemblage_2unt_3m(w):
    """
    TThe assemblage created for Charlie after Alice and Bob each performs three Pauli spin measurements
    (X, Y and Z) on her/his particle of the three-qubit GHZ state.
    """
    Pauli_x = np.array([[0, 1], [1, 0]])
    Pauli_y = np.array([[0, -1j], [1j, 0]])
    Pauli_z = np.array([[1, 0], [0, -1]])
    sigma = np.zeros((2, 2, 2, 2, 3, 3), dtype=complex)
    diagonal = np.diag([(1 - w) / 8, (1 - w) / 8])
    Pauli = [Pauli_x, Pauli_y, Pauli_z]
    for x in range(3):
        for a in range(2):
            for y in range(3):
                for b in range(2):
                    projector_A = (np.eye(2) + ((-1) ** a) * Pauli[x]) / 2
                    projector_B = (np.eye(2) + ((-1) ** b) * Pauli[y]) / 2
                    sigma[:, :, a, b, x, y] = diagonal + (w / 2) * np.array([
                        [projector_A[0, 0] * projector_B[0, 0], projector_A[1, 0] * projector_B[1, 0]],
                        [projector_A[0, 1] * projector_B[0, 1], projector_A[1, 1] * projector_B[1, 1]]
                    ])
    return tf.cast(sigma,'complex64')

File: Bounds for steerability/3_qubit_GHZ_states_2unt/targets.m
classdef targets
    methods(Static)
        % `targets` generates the target assemblages
        % Input:
        %   v - visibilitiy
        % Output:
        %   sigma - the target assemblages

        % requires: nothing
        % author: Mengyan Li(mylmengyanli@gmail.com)
        % last updated: February 24, 2025
        
        function sigma = GHZ_assemblage_2unt_2m(v)
            Pauli_x = [0, 1; 1, 0];
            Pauli_z = [1, 0; 0, -1];
            sigma = zeros(2, 2, 2, 2, 2, 2);
            diagonal = diag([(1-v)/8, (1-v)/8]);
            Pauli = {Pauli_x, Pauli_z};
            for x = 1:2
                for a = 1:2
                    for y = 1:2
                        for b = 1:2
                            projector_A=(eye(2)+(-1)^a*Pauli{x})/2;
                            projector_B=(eye(2)+(-1)^b*Pauli{y})/2;
                            sigma(:,:,a,b,x,y) = diagonal + v/2 *...
                            [projector_A(1,1)*projector_B(1,1) projector_A(2,1)*projector_B(2,1);
                             projector_A(1,2)*projector_B(1,2) projector_A(2,2)*projector_B(2,2)];
                        end
                    end
                end
            end
        end
        function sigma = GHZ_assemblage_2unt_3m(v)
            Pauli_x = [0, 1; 1, 0];
            Pauli_y = [0, -1i; 1i, 0];
            Pauli_z = [1, 0; 0, -1];
            sigma = zeros(2, 2, 2, 2, 3, 3);
            diagonal = diag([(1-v)/8, (1-v)/8]);
            Pauli = {Pauli_x, Pauli_y, Pauli_z};
            for x = 1:3
                for a = 1:2
                    for y = 1:3
                        for b = 1:2
                            projector_A=(eye(2)+(-1)^a*Pauli{x})/2;
                            projector_B=(eye(2)+(-1)^b*Pauli{y})/2;
                            sigma(:,:,a,b,x,y) = diagonal + v/2 *...
                            [projector_A(1,1)*projector_B(1,1) projector_A(2,1)*projector_B(2,1);
                             projector_A(1,2)*projector_B(1,2) projector_A(2,2)*projector_B(2,2)];
                        end
                    end
                end
            end
        end
    end
end